dask_expr/_groupby.py

```diff
@@ -394,11 +421,28 @@ class GroupByReduction(Reduction):
     def _chunk_cls_args(self):
         return [self.by]
 
+    @functools.cached_property
+    def _by_meta(self):
+        if isinstance(self.by, Expr):
+            return meta_nonempty(self.by._meta)
+        elif is_scalar(self.by):
+            return self.by
+        else:
+            return [
+                meta_nonempty(x._meta) if isinstance(x, Expr) else x for x in self.by
+            ]
+
+    @functools.cached_property
+    def _by_columns(self):
+        if isinstance(self.by, Expr):
+            return []
+        else:
+            return [x for x in self.by if not isinstance(x, Expr)]
+
```

Comment:
No, that's not groupby specific, so that's not the place for this logic.

The groupby structure needs a refactor anyway to make this more consistent but that's something for a follow up if the actual implementation is ironed out

Reply:
I rewrote parts of the implementation, still not happy with it, but less duplicated code now